File: Backend/NotesApi/NotesApi/Services/NoteService.cs
using NotesApi.Models;
using NotesApi.Repositories;

namespace NotesApi.Services
{
    public class NoteService : INoteService
    {
        private readonly INoteRepository _noteRepository;

        public NoteService(INoteRepository noteRepository)
        {
            _noteRepository = noteRepository;
        }

        public async Task<IEnumerable<Note>> GetUserNotes(int userId)
        {
            return await _noteRepository.GetUserNotes(userId);
        }

        public async Task<Note> GetNote(int noteId, int userId)
        {
            return await _noteRepository.GetById(noteId, userId);
        }

        public async Task<Note> SaveNote(NoteSaveDto noteDto, int userId)
        {
            return await _noteRepository.Save(noteDto, userId);
        }

        public async Task<bool> DeleteNote(int noteId, int userId)
        {
            return await _noteRepository.Delete(noteId, userId);
        }

        public async Task<IEnumerable<Note>> SearchNotes(int userId, string searchTerm)
        {
            return await _noteRepository.Search(userId, searchTerm);
        }
    }
}

File: Frontend/src/services/noteService.ts
import api from '../utils/fetchApi';
import type { Note, NoteSaveDto } from '../types/noteTypes';

class NoteService {
  private static instance: NoteService;

  private constructor() {}

  public static getInstance(): NoteService {
    if (!NoteService.instance) {
      NoteService.instance = new NoteService();
    }
    return NoteService.instance;
  }

  public async getNotes(): Promise<Note[]> {
    return api.get('/api/notes');
  }

  public async getNote(id: number): Promise<Note> {
    return api.get(`/api/notes/${id}`);
  }

  // New save method for create or update
  public async saveNote(noteDto: NoteSaveDto): Promise<Note> {
    return api.post('/api/notes/save', noteDto);
  }

  public async deleteNote(id: number): Promise<void> {
    return api.delete(`/api/notes/${id}`);
  }

  public async searchNotes(searchTerm: string): Promise<Note[]> {
    return api.get('/api/notes/search', { params: { searchTerm } });
  }
}

export const noteService = NoteService.getInstance();